### PR TITLE
[CELEBORN-1809][CIP-14] Add partitionLocation to cppClient

### PR DESCRIPTION
**Pull request by HolyLow**
### What changes were proposed in this pull request?
This PR adds PartitionLocation to cppClient, which is the component of protocol module.

### Why are the changes needed?
To support communication message of PartitionLocation.

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Compilation and UTs.

### CODE REVIEW

**Comment by HolyLow** [comment]
@FMX Could you kindly review this PR? Thanks a lot.

Any suggestion is welcome.

Also note that the latest java code uses packedPb to transfer the PartitionLocation message, but the code should work with other messages, so we leave the packedPb translation to later commits.

File: cpp/celeborn/protocol/PartitionLocation.h [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+  static const int HDFS_MASK = 0b100;
+  static const int OSS_MASK = 0b1000;
+  static const int ALL_TYPES_AVAILABLE_MASK = 0;
+
+  Type type{MEMORY};
+  std::string mountPoint;
+  // if a file is committed, field "finalResult" will be true
+  bool finalResult{false};
+  std::string filePath;
+  int availableStorageTypes{0};
```
Comment (RexXiong) [inline cpp/celeborn/protocol/PartitionLocation.h:52]: There are two new field fileSize/chunkOffsets added recently

**Review by RexXiong — COMMENTED** [review]
LGTM， expect a nit

File: cpp/celeborn/protocol/PartitionLocation.h (hunk 1 above)
Comment (HolyLow) [reply cpp/celeborn/protocol/PartitionLocation.h:52]: Which file are these new fields added in CelebornJava? I failed to find these fields in PartitionLocation.java or StorageInfo.java.

Comment (RexXiong) [reply cpp/celeborn/protocol/PartitionLocation.h:54]: Sorry, forget this, it's my mistake.

**Review by FMX — APPROVED** [review]
LGTM. Thanks. Merged into main(v0.6.0).

**Pull request closed without merging**